tests/tracing/test_provider.py

```diff
@@ -9,42 +10,63 @@
     _TRACE_BUFFER,
     InferenceTableSpanExporter,
 )
-from mlflow.tracing.export.mlflow import MlflowSpanExporter
 from mlflow.tracing.fluent import TRACE_BUFFER
 from mlflow.tracing.processor.inference_table import InferenceTableSpanProcessor
-from mlflow.tracing.processor.mlflow import MlflowSpanProcessor
 from mlflow.tracing.provider import (
-    _TRACER_PROVIDER_INITIALIZED,
     _get_tracer,
     _is_enabled,
+    _setup_tracer_provider,
+    reset_tracer_setup,
+    start_span_in_context,
     trace_disabled,
 )
 
 
-# Mock client getter just to count the number of calls
-def test_tracer_provider_singleton():
-    # Reset the Once object as there might be other tests that have already initialized it
-    _TRACER_PROVIDER_INITIALIZED._done = False
-    _get_tracer("module_1")
-    assert _TRACER_PROVIDER_INITIALIZED._done is True
+@pytest.fixture
+def mock_setup_tracer_provider():
+    # To count the number of times _setup_tracer_provider is called
+    with mock.patch(
+        "mlflow.tracing.provider._setup_tracer_provider", side_effect=_setup_tracer_provider
+    ) as setup_mock:
+        yield setup_mock
 
-    # Trace provider should be identical for different moments in time
-    tracer_provider_1 = trace.get_tracer_provider()
-    tracer_provider_2 = trace.get_tracer_provider()
-    assert tracer_provider_1 is tracer_provider_2
 
+def test_tracer_provider_initialized_once(mock_setup_tracer_provider):
+    assert mock_setup_tracer_provider.call_count == 0
+    start_span_in_context("test1")
+    assert mock_setup_tracer_provider.call_count == 1
+
+    start_span_in_context("test_2")
+    start_span_in_context("test_3")
+    assert mock_setup_tracer_provider.call_count == 1
+
+    # Thread safety
+    with ThreadPoolExecutor(max_workers=2) as executor:
+        executor.map(start_span_in_context, ["test_4", "test_5"])
+    assert mock_setup_tracer_provider.call_count == 1
 
-def test_span_processor_and_exporter_default():
-    _TRACER_PROVIDER_INITIALIZED._done = False
-    tracer = _get_tracer("test")
-    processors = tracer.span_processor._span_processors
-    assert len(processors) == 1
-    assert isinstance(processors[0], MlflowSpanProcessor)
-    assert isinstance(processors[0].span_exporter, MlflowSpanExporter)
+
+def test_reset_tracer_setup(mock_setup_tracer_provider):
+    assert mock_setup_tracer_provider.call_count == 0
+
+    start_span_in_context("test1")
+    assert mock_setup_tracer_provider.call_count == 1
+
+    reset_tracer_setup()
+    assert mock_setup_tracer_provider.call_count == 2
+
+    start_span_in_context("test2")
+    assert mock_setup_tracer_provider.call_count == 3
+    mock_setup_tracer_provider.assert_has_calls(
+        [
+            mock.call(),
+            mock.call(disabled=True),
+            mock.call(),
+        ]
+    )
```

Comment:
```suggestion
    assert mock_setup_tracer_provider.mock_calls == (
        [
            mock.call(),
            mock.call(disabled=True),
            mock.call(),
        ]
    )
```

Reply:
@harupy What is the differences between two?

Reply:
@B-Step62 Good question.

```python
from unittest import mock


def test_a():
    m = mock.MagicMock()
    m(1)
    m(2)
    m(3)
    m(4)
    m.assert_has_calls(
        [
            mock.call(1),
            mock.call(2),
            mock.call(3),
        ]
    )  # succeeds
    assert m.mock_calls == [
        mock.call(1),
        mock.call(2),
        mock.call(3),
    ]  # fails

```

Reply:
Ooh I thought `assert_has_calls` is exact match. Thank you for the education!!😄